zerver/models.py

```diff
@@ -317,14 +329,30 @@ class GetRealmByDomainException(Exception):
 
 def get_realm_by_email_domain(email):
     # type: (Text) -> Optional[Realm]
+    """The longest subdomain match takes precedent, followed by
+    whether subdomains_allowed is False.  So if we have RealmAlias
+    columns (realm1, example.com, False), (realm2, zulip.example.com,
+    True), (realm3, example.com, True): example.com would match realm1,
+    zulip.example.com would match realm2 and test.example.com would match
+    realm3."""
+
     if settings.REALMS_HAVE_SUBDOMAINS:
         raise GetRealmByDomainException(
             "Cannot get realm from email domain when settings.REALMS_HAVE_SUBDOMAINS = True")
-    try:
-        alias = RealmAlias.objects.select_related('realm').get(domain = email_to_domain(email))
+    domain = email_to_domain(email)
+    query = RealmAlias.objects.select_related('realm')
+    alias = query.filter(domain=domain, subdomains_allowed=False).first() or \
+        query.filter(domain=domain, subdomains_allowed=True).first()
+    if alias:
         return alias.realm
-    except RealmAlias.DoesNotExist:
-        return None
+    else:
+        query = query.filter(subdomains_allowed=True)
+        while domain:
```

Comment:
yeah, zulip is in the "Be careful, since lots of other things are false as well" camp :). In general we avoid the `if not X` idioms in favor of type-specific checks.